### PR TITLE
Adding staging for docker image to support including external components

### PR DESCRIPTION
**Pull request by bettinaheim**
This PR 
- Supersedes https://github.com/NVIDIA/cuda-quantum/pull/285 by refactoring the ext dev dependencies such that the devdeps environment and the release environment both have the same installation of OpenMPI and its runtime dependencies. We might want to change that in the future if it becomes necessary to have additional OpenMPI related dependencies for development, but as it is now it seems we need the dependencies that we use for development also at runtime.
- Deploys a cuda-quantum-dev image that can be used to build additional CUDA Quantum components before creating a release using cudaq.Dockerfile.
- Pushes a commit to the bot/data branch that indicates when a new cuda-quantum-dev image has been pushed so mirrors can pick up on that. 

A manual run of the deployment workflow that will be launched upon merges to main and release branches: https://github.com/NVIDIA/cuda-quantum/actions/runs/5382854071

The idea is that the deployment workflow on GitHub pushes a commit to the bot/data branch that contains the image hash to download the freshly deployed cuda-quantum-dev image. A gitlab workflow (must be checked in [here](https://github.com/NVIDIA/cuda-quantum/blob/bot/data/.gitlab/workflows.yml)) then triggers a pipeline on the GitLab mirror that uses that image to build and deploy additional components. 
A run of a first draft for a pipeline to build the cuquantum_mgpu target can be found [here](https://gitlab-master.nvidia.com/nvhpc/cuda-quantum/-/jobs/63180978). It currently produces the needed components as artifacts that could simply be copied into the final image.

There are two options from there: 
Option 1: make it possible for the GitLab mirror to push packages to GHCR (not currently possible)
Option 2: dump the binaries for the additional components somewhere (a release on GitHub?) and trigger a GitHub workflow to  build and deploy packages versions that include these components.

I'll set one of that up as a separate work item - this PR is getting big enough.

### CODE REVIEW

File: .github/workflows/build_packages.yml [hunk 1]
```diff
@@ -242,20 +170,153 @@ jobs:
           task_name: html
           source_files: ${{ join(fromJSON(steps.docs_build.outputs.json).html_files, ' ') }}
 
+  cudaqdev_deployment:
+    name: Deploy build
+    id: cudaqdev_deployment
```
Comment (bettinaheim) [inline .github/workflows/build_packages.yml:175]:
```suggestion
```

File: .github/workflows/build_packages.yml [hunk 2]
```diff
@@ -242,20 +170,152 @@ jobs:
...
+      tar_cache: ${{ steps.cudaq_image.outputs.tar_cache }}
+      tar_archive: ${{ steps.cudaq_image.outputs.tar_archive }}
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3
+
+      - name: Restore release build
+        id: restore
+        if: needs.cudaqdev_deployment.outcome == 'skipped'
```
Comment (bettinaheim) [inline .github/workflows/build_packages.yml:239]:
```suggestion
        if: needs.cudaqdev_deployment.result == 'skipped'
```

File: .github/workflows/build_packages.yml [hunk 3]
```diff
@@ -242,20 +177,201 @@ jobs:
...
+
+      - name: OpenMPI installation
+        id: open_mpi
+        run: |
+          tar_archive=/tmp/open-mpi.tar
+          docker pull ${{ inputs.ompidev_image }}
+
+          docker build -t open-mpi:local -f docker/build/devdeps.ext.Dockerfile . \
+            --build-arg ompidev_image=${{ inputs.ompidev_image }} \
+            --build-arg cuda_packages=cuda-runtime-11-8
```
Comment (bettinaheim) [inline .github/workflows/build_packages.yml:243]:
```suggestion
            --build-arg cuda_packages=cuda-runtime-11-8 \
```

File: .github/workflows/build_packages.yml [hunk 4]
```diff
@@ -242,20 +177,198 @@ jobs:
...
+            cudaqdev_image=${{ needs.cudaqdev_deployment.outputs.image_hash }}
+          fi
+
+          docker build -t $image_name:$image_tag -f docker/release/cudaq.Dockerfile . \
+            --build-arg cudaqdev_image=$cudaqdev_image \
+            --build-arg base_image=$base_image \
+            --build-arg release_version=$image_tag
+
+          rm -rf "${{ needs.cudaqdev_image.outputs.tar_archive }}"
+          docker image rm ${image_name}-dev:$image_tag
```
Comment (bettinaheim) [inline .github/workflows/build_packages.yml:318]:
```suggestion
          docker image rm $cudaqdev_image
```

File: .github/workflows/build_packages.yml [hunk 5]
```diff
@@ -242,19 +177,205 @@ jobs:
...
+        uses: actions/cache/save@v3
+        with:
+          path: ${{ steps.open_mpi.outputs.tar_archive }}
+          key: ${{ steps.open_mpi.outputs.tar_cache }}
+
+# FIXME: ONLY IF WE ARE NOT DEPLOYING FROM ELSEWHERE
+  cudaq_image:
+    name: Docker image
+    needs: [metadata, ompi_image, cudaqdev_image, cudaqdev_deployment]
+    # Force this job to run even when some of the dependencies (needs) above are skipped.
```
Comment (schweitzpgi) [inline .github/workflows/build_packages.yml:260]:
```suggestion
    # Force this job to run even when some of the dependencies above are skipped.
```

File: .github/workflows/build_packages.yml [hunk 6]
```diff
@@ -242,19 +177,205 @@ jobs:
...
+        run: |
+          if [ -f /tmp/validation.out ]; then
+            echo "## Validation" >> $GITHUB_STEP_SUMMARY
+            echo "The validation of the cuda-quantum image produced the following output:" >> $GITHUB_STEP_SUMMARY
+            echo '```text' >> $GITHUB_STEP_SUMMARY
+            cat /tmp/validation.out >> $GITHUB_STEP_SUMMARY
+            echo '```' >> $GITHUB_STEP_SUMMARY
+          fi
+
+# FIXME: ONLY IF WE ARE NOT DEPLOYING FROM ELSEWHERE
```
Comment (schweitzpgi) [inline .github/workflows/build_packages.yml:354]: What does deploying from elsewhere mean?

**Review by schweitzpgi — APPROVED** [review]

File: .github/workflows/build_packages.yml (hunk 6 above)
Comment (bettinaheim) [reply .github/workflows/build_packages.yml:354]: Ah, let me remove that comment. This was written with Option 1 in the PR description in mind. However, I've decided to defer that to a separate PR, to keep PRs smaller. Also, I think Option 2 might be preferable.

Comment (bettinaheim) [inline .github/workflows/build_packages.yml:354]:
```suggestion
```

File: .github/workflows/build_packages.yml [hunk 7]
```diff
@@ -242,19 +177,205 @@ jobs:
...
+          echo "tar_cache=tar-ompi-${{ github.sha }}" >> $GITHUB_OUTPUT
+          echo "tar_archive=$tar_archive" >> $GITHUB_OUTPUT
+
+      - name: Cache cuda-quantum image
+        uses: actions/cache/save@v3
+        with:
+          path: ${{ steps.open_mpi.outputs.tar_archive }}
+          key: ${{ steps.open_mpi.outputs.tar_cache }}
+
+# FIXME: ONLY IF WE ARE NOT DEPLOYING FROM ELSEWHERE
```
Comment (bettinaheim) [inline .github/workflows/build_packages.yml:256]:
```suggestion
```

**Pull request merged**